File: ObscuritasMediaManager.Frontend/data/session.js
import { FallbackAudio } from '../native-components/fallback-audio/fallback-audio.js';
import { InstrumentModel, MediaModel } from '../obscuritas-media-manager-backend-client.js';
import { MediaService, MusicService } from '../services/backend.services.js';
import { Observable } from './observable.js';

export const Session = {
    /** @type {Observable<string>} */ currentPage: new Observable(''),
    /** @type {Observable<MediaModel[]>} */ mediaList: new Observable([]),
    /** @type {Observable<InstrumentModel[]>} */ instruments: new Observable([]),

    /** @type {FallbackAudio} */ Audio: new FallbackAudio(),

    initialized: false,
    initialize: async () => {
        Session.initialized = false;

        try {
            Session.mediaList.next(await MediaService.getAll());
        } catch (err) {
            console.error(err);
        }

        try {
            Session.instruments.next(await MusicService.getInstruments());
        } catch (err) {
            console.error(err);
        }

        Session.initialized = true;
    },
};

await Session.initialize();

File: ObscuritasMediaManager.Client/ObscuritasMediaManager.Client/Data/Session.cs
using Microsoft.EntityFrameworkCore;
using NAudio.Wave;
using ObscuritasMediaManager.Backend.DataRepositories;
using System;
using System.Linq;

namespace ObscuritasMediaManager.Client.Data;

public class Session(MediaRepository MediaRepository, MusicRepository MusicRepository, UserRepository UserRepository)
{
    public  Observable<string> currentPage { get; set; } = new(string.Empty);
    public  Observable<List<MediaModel>>  mediaList { get; set; } = new(new());
    public  Observable<List<InstrumentModel>> instruments { get; set; } = new(new());
    public  Observable<UserSettingsModel> UserSettings { get; set; } = new(new());
    public  WaveOutEvent Audio { get; set; } = new();
    public  bool initialized { get; set; } = false;

    public async Task InitializeAsync()
    {
        initialized = false;

        try
        {
            mediaList.Next(await MediaRepository.GetAll().ToListAsync());
        }
        catch (Exception err)
        {
            Log.Error(err.ToString());
        }

        try
        {
            instruments.Next((await MusicRepository.GetInstrumentsAsync()).ToList());
        }
        catch (Exception err)
        {
            Log.Error(err.ToString());
        }

        try
        {
            var credentials = await Credentials.FromLocalFileAsync();
            var user = await UserRepository.LogonAsync(credentials.Username, credentials.Password);
            UserSettings.Next(await UserRepository.GetSettingsAsync(user.Id));
        }
        catch (UnauthorizedAccessException)
        {
            throw;
        }
        catch (Exception err)
        {
            Log.Error(err.ToString());
        }

        initialized = true;
    }
}
